---
html_document:
output: html_document
author: Yuyao Song song-yy17@mails.tsinghua.edu.cn
title: "ScfindME: mining alternative splicing patterns in single-cell atlases"
df_print: paged
---
# Install
## Install scfindME from source - mergedWhippet version

```{r load scfindME, warning=FALSE}
#devtools::load_all("~/scfindME")
devtools::install_github("hemberg-lab/scfindME", ref = "mergedWhippet")
library(scfindME)
```

## Build original matrix from scratch
```{r Build matrix from scratch, eval=FALSE}
mo <- buildMatrix.original("/lustre/scratch117/cellgen/team218/gp7/Micro-exons/Runs/Paper/single_cell/MicroExonator/Whippet/Quant/Single_Cell/Pseudo_bulks/pseudo_bulks.psi.tsv")
```

## Use pseudobulk Whippet quant output to build an original node-to-cell Psi matrix
```{r Load pre-built example data}
# load pre-built original matrix
mo <- readRDS("~/mergedWhippet.matrix.original.rds")
mo[1:5, 1:5]

# load metadata, 
metadata <- readRDS("~/lustre/scratch117/cellgen/team218/ys6/pseudobulk_meta.rds")
metadata[1:5, ]
```

## Scale the original matrix
```{r Scale by Z score and diff}
mo.z <- scaleMatrix.z(mo)
mo.diff <- scaleMatrix.diff(mo)
mo.z[1:5, 1:5]
mo.diff[1:5, 1:5]
```

## Build input matrix objects to be further compressed to an index
```{r Scale above and below}

# We use the above matrix to encode splicing events, that is the inclusion of a node in a cell have a PSI value above tissue mean
m.above.z <- buildMatrix.above(mo.z)
m.above.diff <- buildMatrix.above(mo.diff)

# Likewise, the below matrix only contains splicing events that have a PSI value below tissue mean
m.below.z <- buildMatrix.below(mo.z)
m.below.diff <- buildMatrix.below(mo.diff)
```

```{r echo=FALSE}
meta <- readr::read_tsv("/lustre/scratch117/cellgen/team218/gp7/Micro-exons/Runs/Paper/single_cell/MicroExonator/Whippet/Quant/Single_Cell/Pseudo_bulks/pseudo_bulk_membership.tsv")
meta <- data.frame(meta)
meta$primary_type <- gsub("_\\d*$", "", meta$pseudo_bulk_ID)
meta.new <- meta[-(340:345), ]
```

## Index building

```{r}
index.above.z <- buildAltSpliceIndex(m.above.z, metadata = metadata, dataset.name = "Tasic_2016", column.label = "primary_type")
index.above.diff <- buildAltSpliceIndex(m.above.diff, metadata = metadata, dataset.name = "Tasic_2016", column.label = "primary_type")
```
```{r}
index.below.z <- buildAltSpliceIndex(m.below.z, metadata = metadata, dataset.name = "Tasic_2016", column.label = "primary_type")
index.below.diff <- buildAltSpliceIndex(m.below.diff, metadata = metadata, dataset.name = "Tasic_2016", column.label = "primary_type")
```

# Add node info and PSI statistics to index@metadata
## node_list
```{r}
# Details for all nodes in Whippet
nodes_info_all <- readRDS("~/nodes_info.rds")

# Select only the nodes present in index
gene.nodes <- buildMatrix.node_list(m.above.diff, nodes_info_all)

# Add node info to index
index.above.z@metadata$node_list <- gene.nodes
index.above.diff@metadata$node_list <- gene.nodes
```


# Store some useful variables
```{r}
node_list <- scfindGenes(index.above.z)
cell_types <- cellTypeNames(index.above.z)
```

# Find the sigature nodes of a cell type
# Use the diff scailing
```{r echo=TRUE}
markers_Ptgs2 <- cellTypeMarkers(index.above.diff, "Tasic_2016.L2/3_Ptgs2")
markers_Ptgs2
```

# Lets examine how this pair of adjacent exon
e.g. Snap25
ENSMUSG00000027273.13_9
ENSMUSG00000027273.13_10

Which cells have higher than average PSI of node 9?
```{r}
findCellTypes(index.above.z, "ENSMUSG00000027273.13_9")
```

Which cells have lower than average PSI of node 9?
```{r}
findCellTypes(index.below.z, "ENSMUSG00000027273.13_9")
```


#hyperQueryCellTypes
```{r}
hyperQueryCellTypesAS(index.above.diff, c("-ENSMUSG00000027273.13_9","ENSMUSG00000027273.13_10"))

nodeDetails(index.above.z, c("-ENSMUSG00000027273.13_9","ENSMUSG00000027273.13_10"))
```

##Session info for reproducibility
```{r}
sessionInfo()
```

Add a new chunk by clicking the *Insert Chunk* button on the toolbar or by pressing *Ctrl+Alt+I*.

When you save the notebook, an HTML file containing the code and output will be saved alongside it (click the *Preview* button or press *Ctrl+Shift+K* to preview the HTML file).

The preview shows you a rendered HTML copy of the contents of the editor. Consequently, unlike *Knit*, *Preview* does not run any R code chunks. Instead, the output of the chunk when it was last run in the editor is displayed.
